utils/src/main/java/software/amazon/awssdk/utils/AttributeMap.java

```diff
@@ -199,17 +266,291 @@ public <T> Builder put(Key<T> key, T value) {
+            this.dependents = new HashMap<>(dependents.size());
+            dependents.forEach((key, values) -> {
+                Set<Value<?>> newValues = new HashSet<>(values.size());
+                this.dependents.put(key, newValues);
+                values.forEach(v -> newValues.add(valueRemapping.apply(v)));
+            });
+        }
+
+        private void addDependency(Value<?> consumer, Key<?> dependency) {
+            dependents.computeIfAbsent(dependency, k -> new HashSet<>()).add(consumer);
+        }
+
+        private Set<Value<?>> dependents(Key<?> key) {
+            return dependents.getOrDefault(key, emptySet());
+        }
+
+        private void invalidateConsumerCaches(Key<?> key) {
+            Set<Value<?>> dependents = dependents(key);
+            if (dependents.isEmpty()) {
+                return;
+            }
+
+            Queue<Value<?>> unloadQueue = new ArrayDeque<>(dependents);
+            dependents.clear();
+            while (!unloadQueue.isEmpty()) {
+                Value<?> toUnload = unloadQueue.poll();
+
+                Set<Value<?>> toUnloadDependents = dependents(toUnload.key());
+                unloadQueue.addAll(toUnloadDependents);
+                toUnloadDependents.clear();
+
+                toUnload.clearCache();
+            }
+        }
+
+        public DependencyGraph copy(Function<Value<?>, Value<?>> valueRemapping) {
+            return new DependencyGraph(dependents, valueRemapping);
+        }
+    }
+
+    /**
+     * A value stored in this attribute map.
+     */
+    private interface Value<T> extends SdkAutoCloseable {
+        /**
+         * The key associated with this value.
+         */
+        Key<T> key();
+
+        /**
+         * Resolve the stored value using the provided value source.
+         */
+        T get(LazyValueSource source);
+
+        /**
+         * Copy this value, so that modifications like {@link #clearCache()} on this object do not affect the copy.
+         */
+        Value<T> copy();
+
+        /**
+         * If this value is cached, clear that cache.
+         */
+        void clearCache();
+    }
+
+    /**
+     * A constant (unchanging) {@link Value}.
+     */
+    private static class ConstantValue<T> implements Value<T> {
+        private final Key<T> key;
+        private final T value;
+
+        private ConstantValue(Key<T> key, T value) {
+            this.key = key;
+            this.value = value;
+        }
+
+        @Override
+        public Key<T> key() {
+            return key;
+        }
+
+        @Override
+        public T get(LazyValueSource source) {
+            return value;
+        }
+
+        @Override
+        public Value<T> copy() {
+            return this;
+        }
+
+        @Override
+        public void clearCache() {
+        }
+
+        @Override
+        public void close() {
+            IoUtils.closeIfCloseable(value, null);
+            shutdownIfExecutorService(value);
```

Comment:
Related to #4649, this won't work in Java 21

Reply:
Fixed in https://github.com/aws/aws-sdk-java-v2/pull/4686